### PR TITLE
include Utils.Shell in Formula

### PR DESCRIPTION
**Pull request by gregory-nisbet**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Comment by gregory-nisbet** [comment]
This is a patch that makes the `shell_profile` method accessible from the `Formula` class again.

**Comment by gregory-nisbet** [comment]
So I tested this locally and was able to install `nvm` with the two-line change in the PR. Adding a test is proving to be more difficult. Since `shell_profile` is part of the public API for a formula, I wanted to add a test to make sure that it's there.

I've tried everything I can think of to write a test for the presence of the `shell_profile` method.

`Formula.respond_to? :shell_profile` does returns false. Calling `Formula.shell_profile` and checking to make sure it returns a string throws an exception. I tried creating a Dummy class and inheriting from Formula, but apparently you can't have a class local to a method in Ruby.

**Comment by mistydemeo** [comment]
> apparently you can't have a class local to a method in Ruby.

You can make anonymous classes using `Class.new`:

```ruby
klass = Class.new(Formula) do
  # ...
end
```

**Comment by gregory-nisbet** [comment]
Okay I am completely lost on how to write a test for this.

With the patch above installation of `nvm` worked locally and I can refer to `shell_profile` in the body of a class inheriting from Formula.

I'm having trouble understanding exactly how including a module from a class works and "scaling down" the logic so I can understand it. I'm not able to reproduce what's happening inside Homebrew in a toy example.

I have a small toy example of a class `include`-ing a module local to another module in another file. If I use `def shell_profile` then `shell_profile` is accessible from the `my_shell_profile` method and does not need to be qualified with a namespace. However, if I define it as `self.shell_profile` then I cannot use the unqualified method and must explicit qualify it with its module.

```
# util.rb
module Wrapper
    module Utils
        # shell_profile (no error)
        # self.shell_profile (NameError)
        def shell_profile()
            return "4"
        end
    end
end
```

```
# test.rb
require_relative "util"

class Formula
    include Wrapper::Utils

    def initialize()
    end

    def my_shell_profile()
        shell_profile
        # must use Wrapper::Utils::shell_profile if
        # shell_profile was defined using "def self.shell_profile"
    end
end

a = Formula.new()

puts a.my_shell_profile()
```

Here is an excerpt from `utils/shell.rb`

```
 71     # return the shell profile file based on users' preferred shell
 72     def self.shell_profile
 73       SHELL_PROFILE_MAP.fetch(preferred_shell, "~/.bash_profile")
 74     end
```

**Comment by mistydemeo** [comment]
I'm going to go ahead and merge this as-is to stem the bleeding.

**Comment by gregory-nisbet** [comment]
sounds good.

**Comment by gregory-nisbet** [comment]
Don't merge this! It didn't actually fix the issue. It only appeared to because brew executes `git checkout master` in the course of installing a module.

**Comment by gregory-nisbet** [comment]
Okay I think I got it now. Formulas relying on `shell_profile` should work without modification.

I added a test making sure that `shell_profile` is in scope inside the body of a method on a class inheriting from Formula.

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -41,6 +42,7 @@
 class Formula
   include FileUtils
   include Utils::Inreplace
+  include Utils::Shell
```
Comment (mistydemeo) [inline Library/Homebrew/formula.rb]: You shouldn't need to do this anymore if you're having `#shell_profile` call to the fully-qualified name.

**Pull request closed without merging**